File: app/src/main/java/com/utynote/activities/MainActivity.java
package com.utynote.activities;

import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.view.GravityCompat;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;

import com.utynote.R;
import com.utynote.app.ContentView;
import com.utynote.components.map.MapFragment;
import com.utynote.components.nearby.NearbyFragment;
import com.utynote.components.search.SearchFragment;
import com.utynote.databinding.MainContentBinding;
import com.utynote.utils.Function;
import com.utynote.widgets.panel.SlidingUpPanelLayout;

import static com.utynote.utils.Preconditions.checkNotNull;

public class MainActivity extends AppCompatActivity implements ContentView,
        NavigationView.OnNavigationItemSelectedListener {

    private MainContentBinding mContentBinding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mContentBinding = DataBindingUtil.setContentView(this, R.layout.main_content);
        setSupportActionBar(mContentBinding.toolbar);

        mContentBinding.navigationView.setNavigationItemSelectedListener(this);

        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(this, mContentBinding.drawerLayout,
                mContentBinding.toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        mContentBinding.drawerLayout.setDrawerListener(toggle);
        toggle.syncState();

        getSupportFragmentManager()
                .beginTransaction()
                .add(R.id.main_content, new MapFragment(), MapFragment.TAG)
                .add(R.id.panel_content, new NearbyFragment(), NearbyFragment.TAG)
                .commit();
    }

    @Override
    public void onBackPressed() {
        if (mContentBinding.drawerLayout.isDrawerOpen(GravityCompat.START)) {
            mContentBinding.drawerLayout.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.nav_notes) {

        } else if (id == R.id.nav_places) {

        } else if (id == R.id.nav_calendar) {

        } else if (id == R.id.nav_settings) {
            startActivity(new Intent(this, SettingsActivity.class));
        } else if (id == R.id.nav_help) {

        }

        mContentBinding.drawerLayout.closeDrawer(GravityCompat.START);
        return true;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_appbar, menu);

        MenuItem searchMenuItem = menu.findItem(R.id.action_search);
        MenuItemCompat.setOnActionExpandListener(searchMenuItem, new MenuItemCompat.OnActionExpandListener() {
            @Override
            public boolean onMenuItemActionExpand(MenuItem item) {
                replaceFragment(SearchFragment.TAG, SearchFragment::new);
                return true;
            }

            @Override
            public boolean onMenuItemActionCollapse(MenuItem item) {
                replaceFragment(NearbyFragment.TAG, NearbyFragment::new);
                return true;
            }
        });

        return true;
    }

    @NonNull
    @Override
    public Toolbar getToolbar() {
        return checkNotNull(mContentBinding.toolbar);
    }

    @NonNull
    @Override
    public SlidingUpPanelLayout getSlidingPanel() {
        return checkNotNull(mContentBinding.slidingLayout);
    }

    private void replaceFragment(@NonNull String tag, Function.ZeroArgs<Fragment> factory) {
        Fragment fragment = getSupportFragmentManager().findFragmentByTag(tag);
        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.panel_content, fragment == null ? factory.call() : fragment, tag)
                .commit();
    }
}

File: app/src/main/java/com/utynote/app/activities/MainActivity.kt
package com.utynote.app.activities

import android.content.Intent
import android.content.res.Resources
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.design.widget.NavigationView
import android.support.v4.app.Fragment
import android.support.v4.view.GravityCompat
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.SearchView
import android.view.Menu
import com.jakewharton.rxbinding2.support.v4.view.RxMenuItemCompat
import com.jakewharton.rxbinding2.support.v7.widget.RxSearchView
import com.jakewharton.rxbinding2.view.MenuItemActionViewCollapseEvent
import com.jakewharton.rxbinding2.view.MenuItemActionViewExpandEvent
import com.sothree.slidinguppanel.SlidingUpPanelLayout
import com.utynote.R
import com.utynote.app.AppRoot
import com.utynote.components.map.MapFragment
import com.utynote.components.nearby.NearbyFragment
import com.utynote.components.search.SearchFragment
import com.utynote.databinding.MainContentBinding
import com.utynote.utils.Fragments
import java.util.concurrent.TimeUnit


class MainActivity : AppCompatActivity() {

    private class PanelConfig (val parallax : Int,
                               val touch: Boolean,
                               val anchor: Float,
                               val state : SlidingUpPanelLayout.PanelState)

    private val navListener = NavigationView.OnNavigationItemSelectedListener {
        when (it.itemId) {
            R.id.nav_notes    -> { }
            R.id.nav_places   -> { }
            R.id.nav_calendar -> { }
            R.id.nav_help     -> { }
            R.id.nav_settings -> startActivity(Intent(this@MainActivity, SettingsActivity::class.java))
        }
        binding!!.drawerLayout.closeDrawer(GravityCompat.START)
        true
    }

    private val fragments: Fragments = Fragments(supportFragmentManager)

    private var panelConfig: PanelConfig = PanelConfig(0, false, 0f, SlidingUpPanelLayout.PanelState.HIDDEN)
    private var binding: MainContentBinding? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = DataBindingUtil.setContentView<MainContentBinding>(this, R.layout.main_content)

        with(binding!!) {
            setSupportActionBar(toolbar)
            val toggle = ActionBarDrawerToggle(this@MainActivity, drawerLayout, toolbar,
                    R.string.navigation_drawer_open,  R.string.navigation_drawer_close)
            navigationView.setNavigationItemSelectedListener(navListener)
            drawerLayout.setDrawerListener(toggle)
            toggle.syncState()
        }

        savedInstanceState ?: fragments
                .addToContent(MapFragment.TAG, { MapFragment() })
                .addToPanel(NearbyFragment.TAG, { NearbyFragment() })
    }

    override fun onAttachFragment(fragment: Fragment?) {
        super.onAttachFragment(fragment)
        val f = fragment!!

        panelConfig = when (f) {
            is SearchFragment -> {
                (application as AppRoot).getSearchComponent().inject(f)
                PanelConfig(
                        parallax = resources.getDimensionPixelSize(R.dimen.search_panel_paralax),
                        touch = false,
                        anchor = calculateRelativeOffset(resources.getDimensionPixelSize(R.dimen.search_panel_anchor)),
                        state = SlidingUpPanelLayout.PanelState.ANCHORED)
            }
            is NearbyFragment -> {
                PanelConfig(
                        parallax = resources.getDimensionPixelSize(R.dimen.nearby_panel_paralax),
                        touch = true,
                        anchor = calculateRelativeOffset(resources.getDimensionPixelSize(R.dimen.nearby_panel_anchor)),
                        state = SlidingUpPanelLayout.PanelState.COLLAPSED)
            }
            else -> return
        }
        configurePanel()
    }

    override fun onResumeFragments() {
        super.onResumeFragments()
        configurePanel()
    }

    override fun onBackPressed() {
        if (binding!!.drawerLayout.isDrawerOpen(GravityCompat.START)) {
            binding!!.drawerLayout.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_appbar, menu)

        val searchMenuItem = menu.findItem(R.id.action_search)

        RxMenuItemCompat.actionViewEvents(searchMenuItem).subscribe { e ->
            when (e) {
                is MenuItemActionViewExpandEvent   -> fragments.replaceInPanel(SearchFragment.TAG, { SearchFragment() })
                is MenuItemActionViewCollapseEvent -> fragments.replaceInPanel(NearbyFragment.TAG, { NearbyFragment() })
            }
        }

        RxSearchView.queryTextChangeEvents(searchMenuItem.actionView as SearchView)
                .filter    { e -> e.queryText().length > 2 }
                .debounce(1, TimeUnit.SECONDS)
                .filter    { fragments.isAttached(SearchFragment.TAG) }
                .subscribe { e -> fragments.find(SearchFragment.TAG, SearchFragment::class.java).onSearchTerm(e.queryText()) }

        return true
    }

    fun calculateRelativeOffset(absoluteOffset: Int) : Float {
        return (Resources.getSystem().displayMetrics.heightPixels - absoluteOffset) /
                Resources.getSystem().displayMetrics.heightPixels.toFloat()
    }

    fun configurePanel() {
        val panel = binding?.slidingLayout
        if (panel != null ) {
            with(panel) {
                setParallaxOffset(panelConfig.parallax)
                isTouchEnabled = panelConfig.touch
                anchorPoint = panelConfig.anchor
                panelState = panelConfig.state
            }
        }
    }
}
